### PR TITLE
Bug 1986338:  Error creating list of resources in Import YAML

### PR DESCRIPTION
**Pull request by cyril-ui-developer**
/assign @spadgett 

### Before:
Screen Shot 2021-08-09 at 10 06 16 AM

### After:
Screen Shot 2021-09-22 at 1 53 43 PM
Screen Shot 2021-09-22 at 1 55 57 PM

### CODE REVIEW

**Review by spadgett — APPROVED** [review]
/lgtm

**Comment by spadgett** [comment]
/lgtm cancel

It looks like we're not blocking when creating resources. We need to await the response of each resource before continuing to the next.

File: frontend/public/components/edit-yaml.jsx [hunk 1]
```diff
@@ -112,6 +113,20 @@ export const EditYAML_ = connect(stateToProps)(
         }
 
         createResources(objs, isDryRun) {
+          // If a namespace is being created, wait for each request synchronously.
```
Comment (spadgett) [inline frontend/public/components/edit-yaml.jsx:116]: I would just always run the requests synchronously. Create requests should be fast, and the list should never be extremely large.

File: frontend/public/components/edit-yaml.jsx [hunk 2]
```diff
@@ -112,6 +113,20 @@ export const EditYAML_ = connect(stateToProps)(
...
+            async () => {
+              const promiseRequests = [];
+
+              objs.reduce(async (previousObj, currentObj) => {
+                await previousObj;
+                await promiseRequests.push(k8sCreate(this.getModel(currentObj), currentObj));
+              }, Promise.resolve([]));
+
+              return await promiseRequests;
+            };
```
Comment (spadgett) [inline frontend/public/components/edit-yaml.jsx:127]: I think you can do something like this if you make `createResources` async

```ts
              const promises = [];
              for (const obj of objs) {
                const request = await k8sCreate(this.getModel(obj), obj);
                promises.push(request);
              }
              
              return promises;
```

Comment (spadgett) [reply frontend/public/components/edit-yaml.jsx:127]: But you'll need error handling as well.

**Comment by TheRealJon** [comment]
@spadgett @cyril-ui-developer I think there might be some other edge cases similar to this one where resources need to be created synchronously and in a specific sequence. For instance, CRDs and instances of those CRDs, or Roles and RoleBindings. Should we try to proactively address as many of these cases as we can here?

**Comment by spadgett** [comment]
> @spadgett @cyril-ui-developer I think there might be some other edge cases similar to this one where resources need to be created synchronously and in a specific sequence. For instance, CRDs and instances of those CRDs, or Roles and RoleBindings. Should we try to proactively address as many of these cases as we can here?

Roles and role bindings should be OK. I believe you can reference a role that doesn't exist yet.

Good call on CRDs and CRs. Good call. We should address that.

We could either remove the dry-run check altogether and just always create resources asynchronously, or we can check for both namespaces and CRDs. It might be safer to remove dry-run.

File: frontend/public/components/edit-yaml.jsx (hunk 2 above)
Comment (TheRealJon) [reply frontend/public/components/edit-yaml.jsx:127]: Is this also code ever executed as written? It looks like an anonymous function that is never invoked.

Comment (cyril-ui-developer) [reply frontend/public/components/edit-yaml.jsx:127]: @TheRealJon An `async` function always returns a promise. For example:
```
async function fn() {
  return 'hello';
}
fn().then(console.log)
// hello
```
So I think the usual parenthesis` (---)() `for executing anonymous function is unnecessary in this case.

Comment (cyril-ui-developer) [reply frontend/public/components/edit-yaml.jsx:127]: > @TheRealJon An `async` function always returns a promise. For example:
> 
> ```
> async function fn() {
>   return 'hello';
> }
> fn().then(console.log)
> // hello
> ```
> 
> So I think the usual parenthesis`(---)()`for executing anonymous function is unnecessary in this case.
Oops! Never mind the example. It is a function definition.

**Review by TheRealJon — CHANGES_REQUESTED** [review]
I'll second the suggestions made by @spadgett and add that I don't believe the original fix is working as intended. Need to fix the unreachable/dead code issue I pointed out.

File: frontend/public/components/edit-yaml.jsx (hunk 2 above)
Comment (TheRealJon) [reply frontend/public/components/edit-yaml.jsx:127]: > @TheRealJon An `async` function always returns a promise. For example:
> 
> ```
> async function fn() {
>   return 'hello';
> }
> fn().then(console.log)
> // hello
> ```
> 
> So I think the usual parenthesis`(---)()`for executing anonymous function is unnecessary in this case.

@cyril-ui-developer I agree with the first part of this. Async functions will return Promises, but they still need to be invoked in order to do so.

**Comment by cyril-ui-developer** [comment]
> I'll second the suggestions made by @spadgett and add that I don't believe the original fix is working as intended. Need to fix the unreachable/dead code issue I pointed out.

@TheRealJon Got it. It is not working as intended. Will address the issues. Thanks.

File: frontend/public/components/edit-yaml.jsx (hunk 2 above)
Comment (cyril-ui-developer) [reply frontend/public/components/edit-yaml.jsx]: > But you'll need error handling as well.

@spadgett The error is handled here. 
https://github.com/openshift/console/blob/master/frontend/public/components/import-yaml-results.tsx#L116

**Comment by TheRealJon** [comment]
@cyril-ui-developer In testing this, I noticed that one failure causes the same error message to be displayed in the "Creation status" of every resource. I used this yaml:
```yaml
kind: Project
apiVersion: project.openshift.io/v1
metadata:
  name:  test
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: example
  namespace: fail
data:
  example.property.1: hello
  example.property.2: world
  example.property.file: |-
    property.1=value-1
    property.2=value-2
    property.3=value-3
---
apiVersion: v1
kind: Pod
metadata:
  name: example
  namespace: test
spec:
  containers:
    - name: httpd
      image: 'image-registry.openshift-image-registry.svc:5000/openshift/httpd:latest'
      ports:
        - containerPort: 8080
```
And here got this result:
screenshot-localhost_9000-2021 10 04-17_24_15

I also noticed that because the ConfigMap failed, the Pod was not created, which was unexpected. I'll wait to do a more detailed review of this PR since it appears that there are some other bugs to be addressed first:

1. Each row of the import results table should accurately report the status of the corresponding API request 
2. If one creation fails, it shouldn't prevent the creation of subsequent valid resources.

**Comment by cyril-ui-developer** [comment]
@TheRealJon Updated the test. PTAL
Thanks.

**Review by TheRealJon — APPROVED** [review]
/lgtm
/approve

**Pull request merged**